### PR TITLE
Implement LWG-3618 Unnecessary `iter_move` for `transform_view::iterator`

### LINKED ISSUE #2554: LWG-3618 Unnecessary `iter_move` for `transform_view::iterator`
LWG-3618 Unnecessary `iter_move` for `transform_view::iterator`

### PR DESCRIPTION
**Pull request by fsb4000**
Fixes #2554

### CODE REVIEW

**Review by CaseyCarter — APPROVED** [review]

File: stl/inc/ranges [hunk 1]
```diff
@@ -1960,7 +1960,7 @@ namespace ranges {
             }
 
             _NODISCARD constexpr decltype(auto) operator*() const
-                _NOEXCEPT_IDL0(noexcept(_STD invoke(*_Parent->_Fun, *_Current))) /* strengthened */ {
+                _NOEXCEPT_IDL0(noexcept(_STD invoke(*_Parent->_Fun, *_Current))) {
```
Comment (StephanTLavavej) [inline stl/inc/ranges:1963]: Previously, the Standard didn't mention `noexcept`, so we were free to partially strengthen it.

Now, the Standard mandates `noexcept(noexcept(invoke(*parent_->fun_, *current_)))`, so partially weakening it with `_NOEXCEPT_IDL0` is nonconformant.

I believe we need to say plain `noexcept` here. I'm not sure whether the function body needs to change. I also believe that the test needs to be updated:
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/tests/std/tests/P0896R4_views_transform/test.cpp#L489-L499

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

File: stl/inc/ranges [hunk 2]
```diff
@@ -1960,9 +1960,13 @@ namespace ranges {
...
             _NODISCARD constexpr decltype(auto) operator*() const
-                _NOEXCEPT_IDL0(noexcept(_STD invoke(*_Parent->_Fun, *_Current))) /* strengthened */ {
+                noexcept(noexcept(_STD invoke(*_Parent->_Fun, *_Current))) {
 #if _ITERATOR_DEBUG_LEVEL != 0
-                _Check_dereference();
+                try {
+                    _Check_dereference();
+                } catch (...) {
+                    _STL_VERIFY(false, "cannot dereference end transform_view iterator");
+                }
```
Comment (StephanTLavavej) [inline stl/inc/ranges:1969]: In product code, we never use `try`/`catch` directly - it always has to be `_TRY_BEGIN`/`_CATCH_ALL`/`_CATCH_END` in order to respect `_HAS_EXCEPTIONS=0`, e.g.:
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/future#L473-L479

However, should we be try-catching at all here? I observe that `filter_view` marks `_Check_dereference()` as `noexcept`. Should we simply do the same for `transform_view`, and allow exceptions emitted from `_RANGES end(_Parent->_Range)` to slam into `noexcept`? The implementations are identical aside from error messages:

https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/ranges#L1634-L1639
https://github.com/microsoft/STL/blob/afe0800cd6bf0d163bf678ad4e1feb3c130d1ec5/stl/inc/ranges#L1912-L1917

**Review by StephanTLavavej — CHANGES_REQUESTED** [review]

**Comment by fsb4000** [comment]
Yes, I agree that `filter_view::_Check_dereference` and `transform_view::_Check_dereference` should be the same.

Also do we still need to mark the LWG with `ranges` label?

**Review by CaseyCarter — APPROVED** [review]

**Review by StephanTLavavej — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks, this looks great. Unfortunately, the CLA bot was stuck, and although it's been fixed, this has affected all PRs which had pushed commits during a certain window. @fsb4000, apologies for the headache, but can you please close this PR and open a new one, which should allow the license/cla check to pass?

**Comment by StephanTLavavej** [comment]
I'm mirroring this to the MSVC-internal repo - please notify me if any further changes are pushed. (I'll handle verifying that this works with the recent LLVM update.)

**Comment by StephanTLavavej** [comment]
I've pushed an update to the libcxx skips, as deref and `iter_move` are now guaranteed noexcept.

**Review by StephanTLavavej — APPROVED** [review]

**Review by CaseyCarter — APPROVED** [review]

**Comment by StephanTLavavej** [comment]
Thanks for transforming the STL to implement this LWG issue! :tada: :joy_cat: :smile_cat:

**Pull request merged**